High-throughput Data
================

High-throughput experiments have measurements from many features that are taken simultaneously. An example is a gene expression microarray, or a genotyping technology. What we get is measurements for several features (e.g. genes) from several experimental units (e.g. different individuals). The way this data is represented in R is in a matrix.

One such example is in the library GSE 5859 subset.

``` r
library(devtools)
install_github("genomicsclass/GSE5859Subset")

library(GSE5859Subset)
data(GSE5859Subset)
```

The dataset has three objects, and the first one contains actual gene expression measurements for each gene (feature). We also have columns representing experimental units (samples). For 12 different individuls we have measured the gene expression for each gene. The dimensions show 8793 features (genes) measured on 24 individuals.

``` r
head(geneExpression)
```

    ##           GSM136508.CEL.gz GSM136530.CEL.gz GSM136517.CEL.gz
    ## 1007_s_at         6.543954         6.401470         6.298943
    ## 1053_at           7.546708         7.263547         7.201699
    ## 117_at            5.402622         5.050546         5.024917
    ## 121_at            7.892544         7.707754         7.461886
    ## 1255_g_at         3.242779         3.222804         3.185605
    ## 1294_at           7.531754         7.090270         7.466018
    ##           GSM136576.CEL.gz GSM136566.CEL.gz GSM136574.CEL.gz
    ## 1007_s_at         6.837899         6.470689         6.450220
    ## 1053_at           7.052761         6.980207         7.096195
    ## 117_at            5.304313         5.214149         5.173731
    ## 121_at            7.558130         7.819013         7.641136
    ## 1255_g_at         3.195363         3.251915         3.324934
    ## 1294_at           7.122145         7.058973         6.992396
    ##           GSM136575.CEL.gz GSM136569.CEL.gz GSM136568.CEL.gz
    ## 1007_s_at         6.052854         6.387026         6.640583
    ## 1053_at           6.983827         7.060558         7.010453
    ## 117_at            5.022882         5.175134         5.281784
    ## 121_at            7.729267         7.700608         7.615513
    ## 1255_g_at         3.088541         3.184015         3.076940
    ## 1294_at           7.112384         7.194791         6.884312
    ##           GSM136559.CEL.gz GSM136565.CEL.gz GSM136573.CEL.gz
    ## 1007_s_at         6.948474         6.778464         6.595414
    ## 1053_at           6.775048         7.063689         6.864693
    ## 117_at            5.309194         5.071376         5.091403
    ## 121_at            7.992304         7.706135         7.808486
    ## 1255_g_at         3.167413         3.492037         3.231536
    ## 1294_at           7.401553         7.478987         7.065355
    ##           GSM136523.CEL.gz GSM136509.CEL.gz GSM136727.CEL.gz
    ## 1007_s_at         6.255549         6.379983         6.133068
    ## 1053_at           7.174769         7.702533         7.280781
    ## 117_at            5.237160         5.398616         5.401876
    ## 121_at            7.574813         7.511478         7.607461
    ## 1255_g_at         3.208304         3.212051         3.225123
    ## 1294_at           7.344407         7.631689         7.018479
    ##           GSM136510.CEL.gz GSM136515.CEL.gz GSM136522.CEL.gz
    ## 1007_s_at         6.502051         6.331567         6.354293
    ## 1053_at           7.302209         7.456509         7.282859
    ## 117_at            5.395087         5.280535         4.986950
    ## 121_at            7.993732         7.632947         7.706585
    ## 1255_g_at         3.440186         3.185090         3.192436
    ## 1294_at           7.478820         7.577597         7.339535
    ##           GSM136507.CEL.gz GSM136524.CEL.gz GSM136514.CEL.gz
    ## 1007_s_at         6.517539         6.156754         6.037871
    ## 1053_at           7.689282         7.491967         7.413133
    ## 117_at            5.562001         5.039387         5.054133
    ## 121_at            7.612557         7.543667         7.507113
    ## 1255_g_at         3.107306         3.128269         3.085953
    ## 1294_at           7.398595         7.359040         7.377372
    ##           GSM136563.CEL.gz GSM136564.CEL.gz GSM136572.CEL.gz
    ## 1007_s_at         6.639091         6.393338         6.469794
    ## 1053_at           7.028731         6.697240         7.092346
    ## 117_at            5.361298         5.218937         5.340272
    ## 121_at            7.798197         7.976375         7.753480
    ## 1255_g_at         3.174794         3.409032         3.274033
    ## 1294_at           7.467240         7.355222         6.910401

``` r
dim(geneExpression)
```

    ## [1] 8793   24

A typical analysis we can perform on a data set like this is to compare two groups. We have samples from one group, and samples from another. How do we know which columns come from which group? We use the sampleInfo object.

``` r
head(sampleInfo)
```

    ##     ethnicity       date         filename group
    ## 107       ASN 2005-06-23 GSM136508.CEL.gz     1
    ## 122       ASN 2005-06-27 GSM136530.CEL.gz     1
    ## 113       ASN 2005-06-27 GSM136517.CEL.gz     1
    ## 163       ASN 2005-10-28 GSM136576.CEL.gz     1
    ## 153       ASN 2005-10-07 GSM136566.CEL.gz     1
    ## 161       ASN 2005-10-07 GSM136574.CEL.gz     1

``` r
dim(sampleInfo)
```

    ## [1] 24  4

This matrix has dimensions 24 by 4. Each row in sampleInfo is representing a column in the gene expression matrix. If we want to know what group each column of gene expression matrix comes from, we look at the group column in sampleInfo. Now we know that the first 12 columns are cases and the last 12 columns are controls.

``` r
sampleInfo$group
```

    ##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0

If we finish our analysis and find some features that are interesting, we want to be able to report on those features (what they are, the biological meaning they have, etc).

``` r
head(rownames(geneExpression))
```

    ## [1] "1007_s_at" "1053_at"   "117_at"    "121_at"    "1255_g_at" "1294_at"

We can see that the row names are not biologically meaningful, they are just an ID used by the manufacturer. But they also provide information about the genes that go with the IDs in the gene annotation table which has one row for each gene. The matrix is ordered to match the measurement table. What it shows is the chromosome this gene is on, the location, and the symbol. So when we are done we can find the symbols for the features that we find interesting, and report them.

``` r
head(geneAnnotation)
```

    ##      PROBEID  CHR     CHRLOC SYMBOL
    ## 1  1007_s_at chr6   30852327   DDR1
    ## 30   1053_at chr7  -73645832   RFC2
    ## 31    117_at chr1  161494036  HSPA6
    ## 32    121_at chr2 -113973574   PAX8
    ## 33 1255_g_at chr6   42123144 GUCA1A
    ## 34   1294_at chr3  -49842638   UBA7

So this is a typical way in which these data are organised. We have a measurement table, usually a matrix, another table giving information about the columns, and a third table that gives you information about the rows.
