R Refresher
================

``` r
library(devtools)
```

    ## Warning: package 'devtools' was built under R version 3.4.4

In this course we use base R rather than dplyr approach. To subset a matrix dat for which the third column is larger than k, you do this:

``` r
#dat[dat[,3] > k,]
```

### Exercise 1

Download and install the GSE5859Subset package then load the data.

``` r
install_github("genomicsclass/GSE5859Subset")
```

    ## Downloading GitHub repo genomicsclass/GSE5859Subset@master
    ## from URL https://api.github.com/repos/genomicsclass/GSE5859Subset/zipball/master

    ## Installing GSE5859Subset

    ## '/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file  \
    ##   --no-environ --no-save --no-restore --quiet CMD INSTALL  \
    ##   '/private/var/folders/ps/cc8h34252m5493xyx_tchmrm0000gn/T/RtmpqcZiEx/devtools5ee050e0e376/genomicsclass-GSE5859Subset-8ada5f4'  \
    ##   --library='/Library/Frameworks/R.framework/Versions/3.4/Resources/library'  \
    ##   --install-tests

    ## 

``` r
library(GSE5859Subset)
data(GSE5859Subset)
```

How many samples were processed on 2005-06-27?

``` r
sampleInfo[sampleInfo[,2] == '2005-06-27',]
```

    ##     ethnicity       date         filename group
    ## 122       ASN 2005-06-27 GSM136530.CEL.gz     1
    ## 113       ASN 2005-06-27 GSM136517.CEL.gz     1
    ## 118       ASN 2005-06-27 GSM136523.CEL.gz     0
    ## 117       ASN 2005-06-27 GSM136522.CEL.gz     0
    ## 119       ASN 2005-06-27 GSM136524.CEL.gz     0

5

### Exercise 2

How many of the genes represented in this particular technology are on chromosome Y?

``` r
nona <- na.omit(geneAnnotation)
nona[nona[,2] == 'chrY',]
```

    ##          PROBEID  CHR    CHRLOC  SYMBOL
    ## 1149   201909_at chrY   2709623  RPS4Y1
    ## 3079 204409_s_at chrY  22737611  EIF1AY
    ## 3550   205000_at chrY  15016019   DDX3Y
    ## 4728   206279_at chrY   7142013    PRKY
    ## 5034   206624_at chrY  14813160   USP9Y
    ## 5149 206700_s_at chrY -21867301   KDM5D
    ## 5210   206769_at chrY  15815447  TMSB4Y
    ## 5353   206922_at chrY -16097652     VCY
    ## 5786 207247_s_at chrY   2803518     ZFY
    ## 6133   207647_at chrY -26191940    CDY1
    ## 6389   207893_at chrY  -2654896     SRY
    ## 6402 207912_s_at chrY -25275502    DAZ1
    ## 6417 207918_s_at chrY   9304564   TSPY1
    ## 6862 208220_x_at chrY  -6733959   AMELY
    ## 6913 208281_x_at chrY -26909216    DAZ3
    ## 6923 208282_x_at chrY  25365604    DAZ2
    ## 6968   208307_at chrY  23673258 RBMY1A1
    ## 7001   208331_at chrY  25130410    BPY2
    ## 7004   208332_at chrY -24217903     PRY
    ## 7015   208339_at chrY -19880860    XKRY
    ## 8856   211149_at chrY -15434914     UTY

21

### Exercise 3

What is the log expression value of the for gene ARPC1A on the one subject that we measured on 2005-06-10?

``` r
symbol <- which(geneAnnotation$SYMBOL=="ARPC1A")
date <- which(sampleInfo$date=="2005-06-10")
geneExpression[symbol,date]
```

    ## [1] 8.233599

### Exercise 4

Use the function apply to find the median value of each column. What is the median value of these values?

``` r
medians <- apply(geneExpression,2,FUN=median)
median(medians)
```

    ## [1] 5.421568

### Exercise 5

Write a function that takes a vector of values e and a binary vector group coding two groups, and returns the p-value from a t-test: t.test( e\[group==1\], e\[group==0\])$p.value.

Now define g to code cases (1) and controls (0) like this g &lt;- factor(sampleInfo$group)

Next use the function apply to run a t-test for each row of geneExpression and obtain the p-value. What is smallest p-value among all these t-tests?

``` r
myfunction1 <- function(e){
  t.test(e[g==1],e[g==0])$p.value
}

g <- factor(sampleInfo$group)

pvalues <- apply(geneExpression,1,FUN=myfunction1)
min(pvalues)
```

    ## [1] 1.406803e-21
